### PR TITLE
feature/call-to-api-slider

### PR DESCRIPTION
**Pull request by PietyE**
**Implement call to api in slider** : ( https://trello.com/c/0FJSHaOO/7-implement-call-to-api-for-slider )

- created endpoints to call api for previous and upcoming launches
- create folder redux into SliderComponent and added reducers,actions,actionsCreators etc for state-management
- created function and implemented into slider to get array of slides with data from api

### CODE REVIEW

**Comment by PietyE** [comment]
Not refactored all structure for now

File: src/components/Slider/services/getSlides.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import {
+  sliderGetData,
+  sliderGetDataError,
+  sliderGetDataSuccess,
+} from '../redux/sliderActionsCreators';
+import { fetchData } from './sliderFetch';
+import { SliderAnyActions } from '../types/SliderTypes';
+
+export const getSlides =
+  (url1: string, url2: string) =>
```
Comment (vovkis13) [inline src/components/Slider/services/getSlides.ts:11]: use more convenient names

File: src/components/Slider/services/getSlides.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  async (dispatch: Dispatch<SliderAnyActions>): Promise<void> => {
+    try {
+      dispatch(sliderGetData());
+      const data = await Promise.all([fetchData(url1), fetchData(url2)]);
+      const slides = data.flatMap((array, index) =>
```
Comment (vovkis13) [inline src/components/Slider/services/getSlides.ts:16]: use more convenient name

File: src/components/Slider/services/getSlides.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        index === 0 ? array.slice(0, 2) : array.slice(0, 1),
+      );
+      dispatch(sliderGetDataSuccess(slides));
+    } catch (e) {
```
Comment (vovkis13) [inline src/components/Slider/services/getSlides.ts:20]: rename `e` to  `error`

File: src/shared/api/constants/endpoints.ts [hunk 4]
```diff
@@ -1,3 +1,5 @@
 export const enum Endpoints {
   BASE_ENDPOINT = 'https://ll.thespacedevs.com/2.2.0',
+  GO_FOR_LAUNCHES_ENDPOINT = 'https://ll.thespacedevs.com/2.2.0/launch/?name=&slug=&rocket__configuration__name=&rocket__configuration__id=&status=1&rocket__spacecraftflight__spacecraft__name=&rocket__spacecraftflight__spacecraft__name__icontains=&rocket__spacecraftflight__spacecraft__id=&rocket__configuration__manufacturer__name=&rocket__configuration__manufacturer__name__icontains=&rocket__configuration__full_name=&rocket__configuration__full_name__icontains=&mission__orbit__name=&mission__orbit__name__icontains=&r_spacex_api_id=&net__gt=&net__lt=&net__gte=&net__lte=&window_start__gt=&window_start__lt=&window_start__gte=&window_start__lte=&window_end__gt=&window_end__lt=&window_end__gte=&window_end__lte=&last_updated__gte=&last_updated__lte=',
```
Comment (vovkis13) [inline src/shared/api/constants/endpoints.ts:3]: why do you insert parameters to the endpoint string?

File: src/components/Slider/types/SliderTypes.ts [hunk 5]
```diff
@@ -1,5 +1,41 @@
...
   name: string;
-  date: string;
-  imageURL: string;
 }
+
+export interface SliderInitialState<T> {
+  slides: Array<T>;
+  isLoading: boolean;
+  error: null | string;
+  activeSlideID: null | string;
```
Comment (vovkis13) [inline src/components/Slider/types/SliderTypes.ts:15]: use camel case (ID --> Id)

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx [hunk 6]
```diff
@@ -1,14 +1,32 @@
...
+export const SliderCardFooter = React.memo(({ id }: SliderCardFooterProps) => {
+  const dispatch = useDispatch<Dispatch<SliderAnyActions>>();
+
+  return (
+    <CardActions className="slider-card-footer">
+      <Button size="medium" color="primary" variant="contained">
+        <Link
+          to="#"
+          className="slider-card-read-more-link"
+          onClick={e => dispatch(sliderSetActiveID(id))}
```
Comment (vovkis13) [inline src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx:23]: use separate handler function

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx [hunk 7]
```diff
@@ -3,12 +3,18 @@ import { Link } from 'react-router-dom';
 import { Button, CardActions } from '@mui/material';
 import './SliderCardFooter.scss';
 
-export const SliderCardFooter = () => (
+interface SliderCardFooterProps {
+  id: string;
+}
```
Comment (topsart) [inline src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx:12]: move interface to common directory of types

File: src/components/Slider/Slider.tsx [hunk 8]
```diff
@@ -8,53 +8,58 @@ import { ArrowDirection } from './constants/constants';
...
-      getSlides(
-        Endpoints.GO_FOR_LAUNCHES_ENDPOINT,
-        Endpoints.PREVIOUS_LAUNCHES_ENDPOINT,
-      ),
-    );
+    getSlides(
+      Endpoints.GO_FOR_LAUNCHES_ENDPOINT,
+      Endpoints.PREVIOUS_LAUNCHES_ENDPOINT,
+    )(dispatch);
   }, []);
```
Comment (topsart) [inline src/components/Slider/Slider.tsx:28]: add dispatch to useEffect dependencies, it allows you to render only at dispatch action

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx [hunk 9]
```diff
@@ -2,19 +2,31 @@ import React from 'react';
...
+export const SliderCardFooter = React.memo(({ id }: SliderCardFooterProps) => {
+  const dispatch = useDispatch<Dispatch<SliderAnyActions>>();
+
+  return (
+    <CardActions className="slider-card-footer">
+      <Button size="medium" color="primary" variant="contained">
+        <Link
+          to="#"
+          className="slider-card-read-more-link"
+          onClick={e => dispatch(sliderSetActiveID(id))}
```
Comment (topsart) [inline src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx:23]: change "e" to more informatively word "event" in handler

File: src/index.tsx [hunk 10]
```diff
@@ -1,16 +1,20 @@
 import React from 'react';
 import { createRoot } from 'react-dom/client';
 import { HashRouter } from 'react-router-dom';
+import { Provider } from 'react-redux';
 import { App } from './App';
+import { store } from './components/Slider/redux/store';
```
Comment (Aleksandra05sh) [inline src/index.tsx:6]: I think const  store should be in folder store, in root of folder srs:
-srs(folder)
   - store(folder)
     - index.ts(file)
       - const store

File: src/index.tsx [hunk 11]
```diff
@@ -1,16 +1,20 @@
...
 const container = document.getElementById('root') as HTMLElement;
 const root = createRoot(container);
 
 root.render(
   <React.StrictMode>
-    <HashRouter>
-      <App />
-    </HashRouter>
+    <Provider store={store}>
+      <HashRouter>
```
Comment (Aleksandra05sh) [inline src/index.tsx:15]: Why do we have <HashRouter> here, if master branch contains <BrowserRouter>

File: src/components/Slider/Slider.tsx [hunk 12]
```diff
@@ -1,38 +1,65 @@
...
+  const dispatch = useDispatch<Dispatch<SliderAnyActions>>();
+
+  useEffect(() => {
+    getSlides(
+      Endpoints.GO_FOR_LAUNCHES_ENDPOINT,
+      Endpoints.PREVIOUS_LAUNCHES_ENDPOINT,
+    )(dispatch);
+  }, []);
+
+  return !error ? (
```
Comment (Aleksandra05sh) [inline src/components/Slider/Slider.tsx:30]: The nesting of ternary operators is not very readable. It makes sense for one of them to move to if condition before the  main return

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx (hunk 7 above)
Comment (PietyE) [reply src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx:12]: why i thought props interface can be into component folder

File: src/index.tsx (hunk 10 above)
Comment (PietyE) [reply src/index.tsx:6]: + , waiting for our redux structure then i will fix it

File: src/shared/api/constants/endpoints.ts (hunk 4 above)
Comment (PietyE) [reply src/shared/api/constants/endpoints.ts:3]: + , fixed

File: src/components/Slider/types/SliderTypes.ts (hunk 5 above)
Comment (PietyE) [reply src/components/Slider/types/SliderTypes.ts:15]: + , fixed

File: src/components/Slider/services/getSlides.ts (hunk 3 above)
Comment (PietyE) [reply src/components/Slider/services/getSlides.ts:20]: fixed

File: src/components/Slider/services/getSlides.ts (hunk 1 above)
Comment (PietyE) [reply src/components/Slider/services/getSlides.ts:11]: renamed

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx (hunk 7 above)
Comment (Aleksandra05sh) [reply src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx:12]: I think all Interfaces should be in types(also for Props)
It is more convenient and  you can extend them or reuse

File: src/components/Slider/types/SliderTypes.ts (hunk 5 above)
Comment (kerrimov) [reply src/components/Slider/types/SliderTypes.ts]: null can't be used

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx (hunk 7 above)
Comment (kerrimov) [reply src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx:12]: Props interfaces can be left in the same file

File: src/components/Slider/redux/sliderActions.ts [hunk 13]
```diff
@@ -0,0 +1,6 @@
+export enum SliderActions {
+  GET_DATA = 'slider/getData',
+  GET_DATA_SUCCESS = 'slider/getDataSuccess',
+  GET_DATA_ERROR = 'slider/getDataError',
+  SET_ACTIVE_ID = 'slider/setActiveID',
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderActions.ts:5]: why do you need `slider/`?

File: src/components/Slider/redux/sliderActionsCreators.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+export const sliderGetData = (): SliderGetData => ({
+  type: SliderActions.GET_DATA,
+});
+export const sliderGetDataSuccess = (
+  slides: Array<Slides>,
+): SliderGetDataSuccess => ({
+  type: SliderActions.GET_DATA_SUCCESS,
+  payload: slides,
+});
+export const sliderGetDataError = (
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderActionsCreators.ts:19]: Leave space between consts

File: src/components/Slider/redux/sliderReducer.ts [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+} from '../types/SliderTypes';
+import { SliderActions } from './sliderActions';
+
+const sliderInitialState: SliderInitialState<Slides> = {
+  slides: [],
+  isLoading: true,
+  error: null,
+  activeSlideId: null,
+};
+export const sliderReducer = (
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderReducer.ts:14]: leave space

File: src/components/Slider/types/SliderTypes.ts [hunk 16]
```diff
@@ -1,5 +1,44 @@
...
+  mission: {
+    description: string;
+  };
 }
+
+export interface SliderInitialState<T> {
+  slides: Array<T>;
+  isLoading: boolean;
+  error: null | string;
+  activeSlideId: null | string;
```
Comment (kerrimov) [inline src/components/Slider/types/SliderTypes.ts:18]: What is T?
null can't be used?

File: src/components/Slider/types/SliderTypes.ts [hunk 17]
```diff
@@ -1,5 +1,44 @@
...
+
+export interface SliderGetDataError extends AnyAction {
+  type: typeof SliderActions.GET_DATA_ERROR;
+  payload: null | string;
+}
+
+export interface SliderSetActiveID extends AnyAction {
+  type: typeof SliderActions.SET_ACTIVE_ID;
+  payload: string;
+}
```
Comment (kerrimov) [inline src/components/Slider/types/SliderTypes.ts:38]: Tou don't need tot extends AnyAction and typeof

File: src/components/Slider/types/SliderTypes.ts [hunk 18]
```diff
@@ -1,5 +1,44 @@
...
+
+export type SliderAnyActions =
+  | SliderGetData
+  | SliderGetDataSuccess
+  | SliderGetDataError
+  | SliderSetActiveID;
```
Comment (kerrimov) [inline src/components/Slider/types/SliderTypes.ts:44]: This one should be removed

File: src/components/Slider/services/sliderFetch.ts [hunk 19]
```diff
@@ -0,0 +1,7 @@
+import axios, { AxiosResponse } from 'axios';
+import { Slides } from '../types/SliderTypes';
+
+export const fetchData = async (url: string): Promise<Array<Slides>> => {
+  const response: AxiosResponse = await axios.get(url);
+  return response.data.results;
```
Comment (kerrimov) [inline src/components/Slider/services/sliderFetch.ts:6]: try catch

File: src/components/Slider/Slider.tsx [hunk 20]
```diff
@@ -1,38 +1,66 @@
...
+              />
+            </SwiperSlide>
+          ))}
+        </Swiper>
+        <SliderArrow direction={ArrowDirection.NEXT} />
+      </Container>
+    )
+  ) : (
+    <p>error</p> //I guess there should be a link to 404 page with error prop
+  );
```
[the remark this answers is not in the document]
Comment (PietyE) [reply src/components/Slider/Slider.tsx]: i left todo

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx [hunk 21]
```diff
@@ -1,14 +1,32 @@
...
-        Read more
-      </Link>
-    </Button>
-  </CardActions>
-);
+interface SliderCardFooterProps {
+  id: string;
+}
+
+export const SliderCardFooter = React.memo(({ id }: SliderCardFooterProps) => {
```
[the remark this answers is not in the document]
Comment (PietyE) [reply src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx]: +

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx [hunk 22]
```diff
@@ -1,14 +1,32 @@
...
+  const dispatch = useDispatch<Dispatch<SliderAnyActions>>();
```
[the remark this answers is not in the document]
Comment (PietyE) [reply src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx]: +

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx (hunk 6 above)
Comment (PietyE) [reply src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx]: +

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx (hunk 9 above)
Comment (PietyE) [reply src/components/Slider/components/SliderCard/components/SliderCardFooter/SliderCardFooter.tsx]: +

File: src/components/Slider/redux/sliderActions.ts (hunk 13 above)
Comment (PietyE) [reply src/components/Slider/redux/sliderActions.ts]: i saw this approach in docs

File: src/shared/api/constants/endpoints.ts [hunk 23]
```diff
@@ -1,3 +1,5 @@
 export const enum Endpoints {
   BASE_ENDPOINT = 'https://ll.thespacedevs.com/2.2.0',
+  GO_FOR_LAUNCHES_ENDPOINT = '/launch/?name=&slug=&rocket__configuration__name=&rocket__configuration__id=&status=1&rocket__spacecraftflight__spacecraft__name=&rocket__spacecraftflight__spacecraft__name__icontains=&rocket__spacecraftflight__spacecraft__id=&rocket__configuration__manufacturer__name=&rocket__configuration__manufacturer__name__icontains=&rocket__configuration__full_name=&rocket__configuration__full_name__icontains=&mission__orbit__name=&mission__orbit__name__icontains=&r_spacex_api_id=&net__gt=&net__lt=&net__gte=&net__lte=&window_start__gt=&window_start__lt=&window_start__gte=&window_start__lte=&window_end__gt=&window_end__lt=&window_end__gte=&window_end__lte=&last_updated__gte=&last_updated__lte=',
```
[the remark this answers is not in the document]
Comment (PietyE) [reply src/shared/api/constants/endpoints.ts:3]: It generated by api service, i tried use other , that is return other launches

File: src/components/Slider/types/SliderTypes.ts (hunk 18 above)
Comment (kerrimov) [reply src/components/Slider/types/SliderTypes.ts:44]: Leaved for reducer

File: src/components/Slider/redux/sliderActions.ts [hunk 24]
```diff
@@ -0,0 +1,6 @@
+export const enum SliderActions {
+  LOAD_SLIDER_DATA = 'loadData',
+  GET_SLIDER_DATA_SUCCESS = 'getDataSuccess',
+  GET_SLIDER_DATA_ERROR = 'getDataError',
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderActions.ts:4]: Use slider in value too

File: src/components/Slider/redux/sliderReducer.ts [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+        ...state,
+        slides: [],
+        isLoading: true,
+        error: null,
+      };
+    case SliderActions.GET_SLIDER_DATA_SUCCESS:
+      return {
+        ...state,
+        slides: payload.flatMap((launchesArray, index: number) =>
+          index === 0 ? launchesArray.slice(0, 2) : launchesArray.slice(0, 1),
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderReducer.ts:30]: what do you want to do? push all launches to store. After that get launches that you need using selector and helpers

File: src/components/Slider/redux/sliderReducer.ts [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+        ),
+        isLoading: false,
+        error: null,
+      };
+    case SliderActions.GET_SLIDER_DATA_ERROR:
+      return {
+        ...state,
+        slides: [],
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderReducer.ts:38]: you don't need to copy initial state

File: src/components/Slider/services/getSlides.ts [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+  sliderGetDataSuccess,
+} from '../redux/sliderActionsCreators';
+import { fetchData } from './sliderFetch';
+import type { SliderAnyActions } from '../types/SliderTypes';
+import { Endpoints } from '../../../shared/api/constants/endpoints';
+
+export const getSlides =
+  (
+    upcomingLaunchesUrl = `${Endpoints.BASE_ENDPOINT}${Endpoints.GO_FOR_LAUNCHES_ENDPOINT}`,
+    previousLaunchesUrl = `${Endpoints.BASE_ENDPOINT}${Endpoints.PREVIOUS_LAUNCHES_ENDPOINT}`,
```
Comment (kerrimov) [inline src/components/Slider/services/getSlides.ts:14]: why did it was passed as an argument?

File: src/components/Slider/services/getSlides.ts [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+  async (dispatch: Dispatch<SliderAnyActions>): Promise<void> => {
+    try {
+      dispatch(sliderGetData());
+      const data = await Promise.all([
+        fetchData(upcomingLaunchesUrl),
+        fetchData(previousLaunchesUrl),
+      ]);
+      dispatch(sliderGetDataSuccess(data));
+    } catch (error) {
+      dispatch(sliderGetDataError(String(error)));
```
Comment (kerrimov) [inline src/components/Slider/services/getSlides.ts:25]: Error is an object, you can't use type conversion

File: src/components/Slider/types/SliderTypes.ts [hunk 29]
```diff
@@ -1,5 +1,44 @@
...
+  error: null | string;
+  activeSlideId: null | string;
+}
+
+export interface SliderGetData extends AnyAction {
+  type: SliderActions.LOAD_SLIDER_DATA;
+}
+
+export interface SliderGetDataSuccess extends AnyAction {
+  type: SliderActions.GET_SLIDER_DATA_SUCCESS;
```
Comment (kerrimov) [inline src/components/Slider/types/SliderTypes.ts:26]: remove extends AnyAction and null

File: src/components/Slider/types/SliderTypes.ts [hunk 30]
```diff
@@ -1,5 +1,44 @@
...
 export interface Slides {
+  image: string;
+  net: string;
+  id: string;
   name: string;
-  date: string;
-  imageURL: string;
+  mission: {
+    description: string;
+  };
```
Comment (kerrimov) [inline src/components/Slider/types/SliderTypes.ts:11]: create interface for mission

**Review by kerrimov — CHANGES_REQUESTED** [review]

File: src/components/Slider/Slider.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (kerrimov) [reply src/components/Slider/Slider.tsx]: Error shouldn't be here

File: src/components/Slider/redux/sliderActions.ts [hunk 31]
```diff
@@ -0,0 +1,8 @@
+export const enum SliderActions {
+  SLIDER_LOAD_DATA = 'sliderLoadData',
+  SLIDER_GET_UPCOMING_LAUNCHES_SUCCESS = 'sliderGetUpcomingLaunchesSuccess',
+  SLIDER_GET_UPCOMING_LAUNCHES_ERROR = 'sliderGetUpcomingLaunchesError',
+  SLIDER_GET_PREVIOUS_LAUNCHES_SUCCESS = 'sliderGetPreviousLaunchesSuccess',
+  SLIDER_GET_PREVIOUS_LAUNCHES_ERROR = 'sliderGetPreviousLaunchesError',
+  SLIDER_SET_ACTIVE_SLIDE_ID = 'sliderSetActiveSlideId',
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderActions.ts:7]: Values should be the same as keys.

File: src/components/Slider/redux/sliderReducer.ts [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+        sliderUpcomingLaunches: [],
+        error: payload,
+        isLoading: false,
+      };
+    case SliderActions.SLIDER_GET_UPCOMING_LAUNCHES_SUCCESS:
+      return {
+        ...state,
+        sliderUpcomingLaunches: payload,
+        error: '',
+        isLoading: false,
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderReducer.ts:34]: don't duplicate initial state

File: src/components/Slider/redux/sliderSelectors.ts [hunk 33]
```diff
@@ -0,0 +1,15 @@
+import { createSelector } from '@reduxjs/toolkit';
+import { StoreState } from '../../../store';
+import type { Slides } from '../types/SliderTypes';
+
+export const getSlides = createSelector(
+  [
+    (state: StoreState) => state.slider.sliderUpcomingLaunches,
+    (state: StoreState) => state.slider.sliderPreviousLaunches,
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderSelectors.ts:8]: move to separate function

File: src/components/Slider/redux/sliderSelectors.ts [hunk 34]
```diff
@@ -0,0 +1,15 @@
...
+  ],
+
+  (sliderUpcomingLaunches, sliderPreviousLaunches): Array<Slides> =>
+    sliderUpcomingLaunches
+      .slice(0, 2)
+      .concat(sliderPreviousLaunches.slice(0, 1)),
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderSelectors.ts:14]: move to helpers

File: src/components/Slider/services/getPreviousLaunches.ts [hunk 35]
```diff
@@ -0,0 +1,33 @@
...
+import type { Dispatch } from 'redux';
+import type {
+  SliderGetPreviousLaunchesError,
+  SliderGetPreviousLaunchesSuccess,
+  SliderLoadData,
+} from '../types/SliderTypes';
+
+export const getPreviousLaunches =
+  (
+    urlPreviousLaunches = `${Endpoints.BASE_ENDPOINT}${Endpoints.PREVIOUS_LAUNCHES_ENDPOINT}`,
```
Comment (kerrimov) [inline src/components/Slider/services/getPreviousLaunches.ts:17]: you don't need this argument, move to function body

File: src/components/Slider/Slider.tsx [hunk 36]
```diff
@@ -1,38 +1,66 @@
...
+export const Slider: React.FC = () => {
+  const slides: Array<Slides> = useSelector(getSlides);
+  const dispatch = useDispatch<Dispatch<SliderAllActions>>();
+
+  useEffect(() => {
+    getPreviousLaunches(
+      `${Endpoints.BASE_ENDPOINT}${Endpoints.PREVIOUS_LAUNCHES_ENDPOINT}`,
+    )(dispatch);
+    getUpcomingLaunches(
+      `${Endpoints.BASE_ENDPOINT}${Endpoints.GO_FOR_LAUNCHES_ENDPOINT}`,
```
Comment (kerrimov) [inline src/components/Slider/Slider.tsx:29]: please read conversation that was before

File: src/components/Slider/components/SliderCard/components/SliderCardFooter/helpers/setSliderActiveId.ts [hunk 37]
```diff
@@ -0,0 +1,9 @@
+import { SliderSetActiveSlideId } from '../../../../../types/SliderTypes';
+import { sliderSetActiveSlideId } from '../../../../../redux/sliderActionsCreators';
+import type { Dispatch } from 'redux';
+
+export const setSliderActiveId =
+  (id: string) =>
+  (dispatch: Dispatch<SliderSetActiveSlideId>): void => {
+    dispatch(sliderSetActiveSlideId(id));
+  };
```
Comment (kerrimov) [inline src/components/Slider/components/SliderCard/components/SliderCardFooter/helpers/setSliderActiveId.ts:9]: curly braces can be removed

File: src/components/Slider/redux/sliderSelectors.ts [hunk 38]
```diff
@@ -0,0 +1,14 @@
...
+
+const sliderUpcomingLaunchesSelector = (state: StoreState) =>
+  state.slider.sliderUpcomingLaunches;
+const sliderPreviousLaunchesSelector = (state: StoreState) =>
+  state.slider.sliderPreviousLaunches;
+
+export const getSlides = createSelector(
+  [sliderUpcomingLaunchesSelector, sliderPreviousLaunchesSelector],
+
+  makeSlides,
```
Comment (kerrimov) [inline src/components/Slider/redux/sliderSelectors.ts:13]: why this space is here?

File: src/components/Slider/services/sliderFetch.ts [hunk 39]
```diff
@@ -0,0 +1,7 @@
+import axios, { AxiosResponse } from 'axios';
+import type { Slides } from '../types/SliderTypes';
+
+export const fetchData = async (url: string): Promise<Array<Slides>> => {
+  const response: AxiosResponse = await axios.get(url);
+  return response.data.results;
+};
```
Comment (kerrimov) [inline src/components/Slider/services/sliderFetch.ts:7]: move to shared/api

File: src/components/Slider/services/getUpcomingLaunches.ts [hunk 40]
```diff
@@ -0,0 +1,30 @@
...
+    >,
+  ): Promise<void> => {
+    try {
+      dispatch(sliderLoadData());
+      const upcomingLaunches = await fetchData(urlUpcomingLaunches);
+      dispatch(sliderGetUpcomingLaunchesSuccess(upcomingLaunches));
+    } catch (error) {
+      dispatch(sliderGetUpcomingLaunchesError((error as Error).message));
+    }
+  };
```
Comment (kerrimov) [inline src/components/Slider/services/getUpcomingLaunches.ts:30]: create common function for `getUpcomingLaunches` and `getPreviousLaunches`

File: src/components/Slider/services/sliderFetch.ts (hunk 39 above)
Comment (kerrimov) [reply src/components/Slider/services/sliderFetch.ts:7]: here is an approach, please discuss it with Volodymyr:
https://github.com/kerrimov/the-ssspace/pull/20/files#diff-ec8cd247324ad888f785b5ff8e963f25aebb4596cb88c30b60ef73eae04210e7

File: src/components/Slider/types/SliderTypes.ts [hunk 41]
```diff
@@ -1,5 +1,59 @@
...
+  id: string;
   name: string;
-  date: string;
-  imageURL: string;
+  mission: LaunchMission;
+}
+
+export interface SliderLoadData extends AnyAction {
+  type: SliderActions.SLIDER_LOAD_DATA;
+}
```
Comment (kerrimov) [inline src/components/Slider/types/SliderTypes.ts:18]: remove AnyAction and don't use destructuring in the reducer. Error with types will be fixed

File: src/components/Slider/types/SliderTypes.ts [hunk 42]
```diff
@@ -1,5 +1,59 @@
...
+  | SliderGetUpcomingLaunchesError
+  | SliderGetPreviousLaunchesSuccess
+  | SliderGetPreviousLaunchesError;
+
+export interface SliderState {
+  sliderUpcomingLaunches: Array<Slides>;
+  sliderPreviousLaunches: Array<Slides>;
+  isLoading: boolean;
+  error: string;
+  activeSlideId: string | null;
```
Comment (kerrimov) [inline src/components/Slider/types/SliderTypes.ts:58]: can't be null

File: src/components/Slider/types/SliderTypes.ts (hunk 41 above)
Comment (PietyE) [reply src/components/Slider/types/SliderTypes.ts:18]: fixed

File: src/components/Slider/types/SliderTypes.ts (hunk 42 above)
Comment (PietyE) [reply src/components/Slider/types/SliderTypes.ts:58]: fixed

File: src/components/Slider/services/getLaunches.ts [hunk 43]
```diff
@@ -0,0 +1,37 @@
...
+        dispatch(sliderGetPreviousLaunchesSuccess(upcomingLaunches));
+      }
+    } catch (error) {
+      if (params === launchesParams.UPCOMING) {
+        dispatch(sliderGetUpcomingLaunchesError((error as Error).message));
+      } else {
+        dispatch(sliderGetPreviousLaunchesError((error as Error).message));
+      }
+    }
+  };
```
Comment (kerrimov) [inline src/components/Slider/services/getLaunches.ts:37]: You don't need params as an argument here and check if it is previous or upcoming. You have added complicated unnecessary logic. Just add two requests side by side. 
Actually, we don't need different actions for upcoming and previous launches, we need all data for the slider, don't we?

Comment (kerrimov) [reply src/components/Slider/services/getLaunches.ts:37]: discussed

File: src/components/Slider/services/getUpcomingLaunches.ts (hunk 40 above)
Comment (kerrimov) [reply src/components/Slider/services/getUpcomingLaunches.ts]: discussed

File: src/shared/api/services/fetchData.ts [hunk 44]
```diff
@@ -0,0 +1,7 @@
+import { axiosInstance } from './axiosInstance';
+import type { AxiosResponse } from 'axios';
+
+export const fetchData = async <T>(url: string): Promise<T> => {
```
Comment (PietyE) [inline src/shared/api/services/fetchData.ts:4]: generic added for my teammates , after invoke this function we can use generic to type result data that we will get and use

File: src/components/Slider/services/sliderFetch.ts (hunk 39 above)
Comment (PietyE) [reply src/components/Slider/services/sliderFetch.ts]: moved to shared and added generic for whole team

**Review by kerrimov — APPROVED** [review]

**Pull request merged**